vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -191,47 +188,62 @@
-     * Supported component types are {@code Byte}, {@code Double}, {@code Float}, {@code Integer}, {@code Long},
-     * {@code Short}, {@code BigInteger} and {@code BigDecimal}.
+     * <p>
+     * The average calculation supports all elements that are of type {@code Number}.
      * <p>
      * Examples:
      * <pre>
      * <code>
-     * List.empty().average()              // = None
-     * List.of(1, 2, 3).average()          // = Some(2.0)
-     * List.of(0.1, 0.2, 0.3).average()    // = Some(0.2)
-     * List.of("apple", "pear").average()  // throws
+     * List.empty().average()                       // = None
+     * List.of(1, 2, 3).average()                   // = Some(2.0)
+     * List.of(1.0, 10e100, 2.0, -10e100).average() // = Some(0.75)
+     * List.of("apple", "pear").average()           // throws
      * </code>
      * </pre>
      *
+     * Please note that Java's {@link DoubleStream#average()} uses the
+     * <a href="https://en.wikipedia.org/wiki/Kahan_summation_algorithm">Kahan summation algorithm</a>
+     * (also known as compensated summation), which has known limitations.
+     * <p>
+     * Vavr uses Neumaier's modification of the Kahan algorithm, which yields better results.
+     *
+     * <pre>{@code
+     * // = Optional(0.0) (wrong)
+     * j.u.s.DoubleStream.of(1.0, 10e100, 2.0, -10e100).average()
+     *
+     * // = Some(0.75) (correct)
+     * List.of(1.0, 10e100, 2.0, -10e100).average()
+     * }</pre>
+     *
      * @return {@code Some(average)} or {@code None}, if there are no elements
      * @throws UnsupportedOperationException if this elements are not numeric
      */
-    @SuppressWarnings({ "unchecked", "OptionalGetWithoutIsPresent" })
     default Option<Double> average() {
-        if (isEmpty()) {
-            return Option.none();
-        } else {
-            final Traversable<?> objects = isTraversableAgain() ? this : toStream();
-            final Object o = objects.head();
-            if (o instanceof Number) {
-                final Traversable<Number> numbers = (Traversable<Number>) objects;
-                final double d;
-                if (o instanceof Integer || o instanceof Long || o instanceof Byte || o instanceof BigInteger || o instanceof Short) {
-                    d = numbers.toJavaStream()
-                            .mapToLong(Number::longValue)
-                            .average()
-                            .getAsDouble();
+        try {
+            @SuppressWarnings("unchecked")
+            final double[] sum = ((Traversable<Number>) this).map(Number::doubleValue)
+                    .foldLeft(new double[4], (acc, d) -> { // acc = [sum, c, count, simpleSum]
+                        final double tmp = acc[0] + d;
+                        acc[1] += (Math.abs(acc[0]) >= Math.abs(d)) ? (acc[0] - tmp) + d : (d - tmp) + acc[0];
+                        acc[0] = tmp;
+                        acc[2]++;
```

Comment:
YIKES!  A side-effect! `acc[2] = acc[2] + 1`

Reply:
😄 don't expect the internal implementation of functional libraries to be side-effect free.

In fact I advocate that mutable state is perfectly ok. In the ideal case it is in a local scope. _Shared_ mutable state is evil!

For example our [Future](https://github.com/vavr-io/vavr/blob/1e6af93c618198aa722714226a682eca13132d64/vavr/src/main/java/io/vavr/concurrent/FutureImpl.java#L84-L104), [Lazy](https://github.com/vavr-io/vavr/blob/1e6af93c618198aa722714226a682eca13132d64/vavr/src/main/java/io/vavr/Lazy.java#L65-L67) and [Iterator](https://github.com/vavr-io/vavr/blob/1e6af93c618198aa722714226a682eca13132d64/vavr/src/main/java/io/vavr/collection/Iterator.java#L46-L48) implementations would not work at all without internal mutable state!

Reply:
I was just being _dogmatic_ for sake of humor